### PR TITLE
Support JsonRPC structure for events - Part II

### PR DESCRIPTION
**Pull request by ksentak**
MF has the ability to trigger events either manually through the API/CLI or automatically through triggers.

In order to send and format these events, MF must keep track of which incoming users are listening. This formatting/searching/etc is currently hard-coded for the FB format. We must open up the existing event functionality to support any JsonRPC structure for events.

### CODE REVIEW

File: server/src/events.mjs [hunk 1]
```diff
@@ -87,70 +100,152 @@ function getRegisteredEventListener(userId, method) {
...
+}
+
+/**
+ * Extracts event data from a given message object based on provided configuration.
+ * @param {object} oMsg - The message object to extract event data from.
+ * @param {object} config - The configuration object for the event data extraction.
+ * @param {boolean} isOn - Whether eventListener enable or disable request
+ * @returns {object | false} - An object containing the extracted event data or false if the extraction fails.
+*/
+function extractEventData(oMsg, config, isOn) {
```
Comment (kschrief) [inline server/src/events.mjs:137]: I'd recommend renaming "isOn" to something a bit more intuitive like "isEnable"

File: server/src/events.mjs [hunk 2]
```diff
@@ -87,70 +100,152 @@ function getRegisteredEventListener(userId, method) {
...
+  const searchRegex = config.searchRegex;
+  const method = config.method || '$.method';
+
+  if (!new RegExp(searchRegex).test(JSON.stringify(oMsg))) {
+    return false;
   }
+
+  const extractedMethod = JSONPath.query(oMsg, method);
+
+  if (extractedMethod.length === 0) {
```
Comment (kschrief) [inline server/src/events.mjs:147]: Can we add a log here to indicate why we're failing?

File: docs/Events.md [hunk 3]
```diff
@@ -1,22 +1,42 @@
...
 
 ### Event
 
+When the registered event occurs, the server sends an event message to the client. Here is an example of an event message:
+
 ```
 {"jsonrpc":"2.0","result":{"state":"inactive","previous":"foreground"}, "id":1}
 ```
+
+## Configuration
```
Comment (kschrief) [inline docs/Events.md:31]: We should add an example flow here. The user will need to know what they can access via handlebars and how. Walking through a flow using the default config in .mf.config.json would do nicely.

File: server/src/events.mjs [hunk 4]
```diff
@@ -87,70 +100,152 @@ function getRegisteredEventListener(userId, method) {
...
+/**
+ * Determines whether the given message is intended to disable an event listener.
+ * A regex pattern and JSON path are provided in config to check whether the event listener is off.
+ * @param {Object} oMsg - The incoming message to check.
+ * @return {boolean} - Returns `true` if the message is intended to disable an event listener, `false` otherwise.
+ */
 function isEventListenerOffMessage(oMsg) {
-  if ( ! isEventListenerMessage(oMsg) ) { return false; }
-  return ( oMsg.params && ! oMsg.params.listen );
+  if (!isEventListenerMessage(oMsg, eventConfig.unRegistrationMessage)) {
```
Comment (kschrief) [inline server/src/events.mjs:203]: This seems overly complex.

I'd say we can make this a one-liner (return isEventListenerMessage(...)) but that also seems overly complex considering we can just use that one-liner where we'd want to call "isEventListenerOffMessage()".

Can this be simplified?

File: server/src/events.mjs [hunk 5]
```diff
@@ -161,20 +256,34 @@ function sendBroadcastEvent(ws, userId, method, result, msg, fSuccess, fErr, fFa
...
+  // Defines the data object that will be inputted into handlebars
+  const templateData = {
+    ...metadata,
+    result: finalResult,
+    resultAsJson: JSON.stringify(finalResult)
   };
-  const eventMessage = JSON.stringify(oEventMessage);
+
+  // If event template config does not exist, use default template which just returns the result
+  const defaultEventTemplate = `{"result":{{{resultAsJson}}}{{~"}}"~}}}`;
```
Comment (kschrief) [inline server/src/events.mjs:283]: We shouldn't do any formatting at all unless a template exists. Just send the exact event payload that was given to the API

File: server/src/messageHandler.mjs [hunk 6]
```diff
@@ -80,18 +83,47 @@ async function handleMessage(message, userId, ws) {
...
+      if (eventConfig.registrationAck) {
+        events.sendEventListenerAck(userId, ws, eventMetadata);
+      }
+  
+      events.registerEventListener(userId, eventMetadata, ws);
+      return;
+    }
+    // If in proxy mode, just register the event listener without sending ack messages or returning
+    else {
+      events.registerEventListener(userId, eventMetadata, ws);
```
Comment (kschrief) [inline server/src/messageHandler.mjs:103]: There's some duplicate code here. We always have to register the listener so we can do that first without putting it in the if/else block.

File: server/src/messageHandler.mjs [hunk 7]
```diff
@@ -80,18 +83,47 @@ async function handleMessage(message, userId, ws) {
...
+      if (eventConfig.unRegistrationAck) {
+        events.sendUnRegistrationAck(userId, ws, eventMetadata);
+      }
+      
+      events.deregisterEventListener(userId, eventMetadata, ws);
+      return;
+    }
+    // If in proxy mode, just deregister the event listener without sending ack messages or returning
+    else {
+      events.deregisterEventListener(userId, eventMetadata, ws);
```
Comment (kschrief) [inline server/src/messageHandler.mjs:125]: See above duplicate code comment

File: server/src/events.mjs (hunk 2 above)
Comment (ksentak) [reply server/src/events.mjs:147]: Added following log

    logger.debug(`Error occurred while extracting event data: No method found in the provided message.`);

File: server/src/events.mjs (hunk 4 above)
Comment (ksentak) [reply server/src/events.mjs]: Kept separate functions so code is easier to read/understand but simplified the functions into one liners

**Review by kschrief — CHANGES_REQUESTED** [review]

File: docs/Events.md [hunk 8]
```diff
@@ -1,22 +1,100 @@
...
+
+```
+// Handlebars config
+{\"result\":{{{resultAsJson}}},\"id\":{{registration.id}},\"jsonrpc\":\"2.0\"}
+
+// Actual result
+{"result": "NEW-DEVICE-NAME-1","id":7,"jsonrpc":"2.0"}
+```
+
+In the Handlebars templates, you can access any data from the event metadata using the dot notation. For instance, you can use {{metadata.params.id}} to access the id from the params object in the metadata.
```
[the remark this answers is not in the document]
Comment (ksentak) [reply docs/Events.md]: In the Handlebars templates, you can access data from the event registration, unregistration, and the event itself, which we collectively refer to as metadata. You can use dot notation to access specific pieces of data. For instance, you can use {{registration.id}} to access the id from the registration object in the metadata. Similarly, you can access information from unregistration and event metadata objects.


Changed to^^

**Review by kschrief — APPROVED** [review]

File: server/src/events.mjs [hunk 9]
```diff
@@ -48,22 +52,31 @@ function logFatalErr() {
...
+  const { registration, method } = metadata;
+
   if (!eventListenerMap[userId]) {
     eventListenerMap[userId] = {};
   }
 
-  if (!eventListenerMap[userId][oMsg.method]) {
-    eventListenerMap[userId][oMsg.method] = { id: oMsg.id, wsArr: [] };
+  if (!eventListenerMap[userId][method]) {
+    // id: registration.id will be removed in part 2
```
Comment (michaelcfine) [inline server/src/events.mjs:70]: Wondering if this represents something left-over / to do since this is the PR for part 2, right?

**Review by michaelcfine — APPROVED** [review]
👍 Looks great. One comment about a comment in the code that looks like maaaybe something didn't get done?.

File: server/src/events.mjs (hunk 9 above)
Comment (ksentak) [reply server/src/events.mjs]: Nice catch, forgot to clean up!

**Review by kschrief — APPROVED** [review]

**Pull request merged**